### PR TITLE
Copy Dockerfiles from private dev repo to devops/

### PR DESCRIPTION
**Pull request by karenc**
- Copy Dockerfiles from private dev repo to devops/

  `devops/Dockerfile`, `devops/base/Dockerfile` and
  `devops/dependencies/Dockerfile` are unchanged and
  `devops/provision/Dockerfile` has been modified to use public repos on
  github instead of private azure ones.
  
  `devops/provision/_config/types_c.h` is required by the provision
  Dockerfile so has also been copied.
  
  `theanorc`, `embed.sh`, `healthcheck.py` and `boto.cfg` in
  `devops/provision/_config/` are required by the ibeis Dockerfile so have
  been copied.  The aws access key id and secret access key have been
  removed from `boto.cfg`.

- Change Dockerfiles to use base images outside of azure

  The docker images on azurecr.io need authentication, so changing them to
  use docker hub or locally built images instead:
  
  - `wildme.azurecr.io/ibeis/base:latest` -> `wildme/ibeis-base:latest`
  - `wildme.azurecr.io/ibeis/dependencies:latest` -> `wildme/ibeis-dependencies:latest`
  - `wildme.azurecr.io/ibeis/provision:latest` -> `wildme/ibeis-provision:latest`

- Add container build script and instructions

- Use the ibeis master branch in Dockerfile

  The "next" branch on the ibeis github repo is now deprecated.  We are
  switching to gitflow and "next" is only around just in case.

- Set LC_ALL and LANG to C.UTF-8 in Dockerfile

  When running `super_setup.py`:
  
  ```
  Traceback (most recent call last):
    File "super_setup.py", line 866, in <module>
      main()
    File "super_setup.py", line 862, in main
      cli_group()
    File "/virtualenv/env3/lib/python3.6/site-packages/click/core.py", line 829, in __call__
      return self.main(*args, **kwargs)
    File "/virtualenv/env3/lib/python3.6/site-packages/click/core.py", line 760, in main
      _verify_python3_env()
    File "/virtualenv/env3/lib/python3.6/site-packages/click/_unicodefun.py", line 130, in _verify_python3_env
      " mitigation steps.{}".format(extra)
  RuntimeError: Click will abort further execution because Python 3 was configured to use ASCII as encoding for the environment. Consult https://click.palletsprojects.com/python3/ for mitigation steps.
  
  This system supports the C.UTF-8 locale which is recommended. You might be able to resolve your issue by exporting the following environment variables:
  
      export LC_ALL=C.UTF-8
      export LANG=C.UTF-8
  ```
  
  So just set the environment variables in the provision Dockerfile.

- Adjust super_setup.py usage in Dockerfile

  The interface for `super_setup.py` has changed so it's just `pull`
  (subcommand) instead of `--pull`:
  
  ```
  Usage: super_setup.py [OPTIONS] COMMAND [ARGS]...
  
  Options:
    -h, --help  Show this message and exit.
  
  Commands:
    check         Check is just a dry run of "ensure".
    develop
    doctest
    ensure        Ensure is the live run of "check".
    ensure_clone
    pull
    status
    versions
  ```

- Temporarily disable super_setup.py in Dockerfiles

  The repos super_setup.py expecting are different from the ones we check
  out, for example "guitool" vs "guitool_ibeis", so just comment out the
  super_setup.py lines in the dockerfiles for now.

- Temporarily install flann v2.0.0 in Dockerfile

  We usually install the "next" branch for all of our github repos.  But
  for flann, the unix_build.sh on the next branch isn't working after
  recent changes, so use the latest tag v2.0.0 instead.
  
  The latest version of flann requires a cmake version 3.13.0 or higher
  but ubuntu 18.04 only has 3.10.2.  We will install a newer cmake version
  in the future.

- Temporarily install hesaff v2.0.0 in Dockerfile

  We usually install the "next" branch for all of our github repos.  But
  for hesaff, the unix_build.sh on the next branch isn't working after
  recent changes, so use the latest tag v2.0.0 instead.
  
  The hesaff "next" branch requires cmake version 3.13.0 or higher but ubuntu
  18.04 only has version 3.10.2.  We are going to install a newer version of
  cmake in the future to fix this.

- Update vtool path to unix_build.sh in Dockerfile

  The path to `unix_build.sh` in vtool has moved in the latest "next"
  branch from the repository root `.` to `./dev`.

- Install scikit-build in provision Dockerfile

  The new `guitool` and `vtool` import `skbuild` from `setup.py` so
  requires `scikit-build` to be installed.

- Activate virtualenv before install in dockerfile

  When installing vtool without activating virtualenv, cmake fails:
  
  ```
      ERROR: Command errored out with exit status 1:
       command: /virtualenv/env3/bin/python3 -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/ibeis/vtool/setup.py'"'"'; __file__='"'"'/ibeis/vtool/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' develop --no-deps
           cwd: /ibeis/vtool/
      Complete output (117 lines):
      Not searching for unused variables given on the command line.
      CMake Error: CMake was unable to find a build program corresponding to "Ninja".  CMAKE_MAKE_PROGRAM is not set.  You probably need to select a different build tool.
      -- Configuring incomplete, errors occurred!
  ...
      ---------------------------
      --------------------------------
      -- Trying "Unix Makefiles" generator - success
      --------------------------------------------------------------------------------
  
      Configuring Project
        Working directory:
          /ibeis/vtool/_skbuild/linux-x86_64-3.6/cmake-build
        Command:
          cmake /ibeis/vtool -G 'Unix Makefiles' -DCMAKE_INSTALL_PREFIX:PATH=/ibeis/vtool/_skbuild/linux-x86_64-3.6/cmake-install -DPYTHON_EXECUTABLE:FILEPATH=/virtualenv/env3/bin/python3 -DPYTHON_VERSION_STRING:STRING=3.6.9 -DPYTHON_INCLUDE_DIR:PATH=/usr/include/python3.6m -DPYTHON_LIBRARY:FILEPATH=/usr/lib/x86_64-linux-gnu/libpython3.6m.so -DSKBUILD:BOOL=TRUE -DCMAKE_MODULE_PATH:PATH=/virtualenv/env3/lib/python3.6/site-packages/skbuild/resources/cmake -DCMAKE_BUILD_TYPE:STRING=Release
  
      Traceback (most recent call last):
  
      ----------------------------------------
    Rolling back uninstall of vtool-ibeis
    Moving to /virtualenv/env3/lib/python3.6/site-packages/vtool_ibeis-2.1.1.dist-info/
     from /virtualenv/env3/lib/python3.6/site-packages/~tool_ibeis-2.1.1.dist-info
    Moving to /virtualenv/env3/lib/python3.6/site-packages/vtool_ibeis/
     from /virtualenv/env3/lib/python3.6/site-packages/~tool_ibeis
  ERROR: Command errored out with exit status 1: /virtualenv/env3/bin/python3 -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/ibeis/vtool/setup.py'"'"'; __file__='"'"'/ibeis/vtool/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' develop --no-deps Check the logs for full command output.
  ```
  
  Activating the virtualenv before installing fixes this problem.

### CODE REVIEW

**Comment by karenc** [comment]
When I was trying to build the provisions Dockerfile, I get this error (which wasn't there yesterday):

```
Step 9/13 : RUN cd /ibeis/utool  && /virtualenv/env3/bin/pip install -e .  && cd /ibeis/ubelt  && /virtualenv/env3/bin/pip install -e .
 ---> Running in 7da00a169b35                        
Obtaining file:///ibeis/utool                        
Requirement already satisfied: six in /virtualenv/env3/lib/python3.6/site-packages (from utool==2.1.4) (1.14.0)
Collecting parse                                     
  Downloading parse-1.15.0.tar.gz (29 kB)            
...
Installing collected packages: parse, pint, smmap, gitdb, gitpython, pytz, babel, humanize, tzlocal, delorean, pandas, utool
  Running setup.py develop for utool
Successfully installed babel-2.8.0 delorean-1.0.0 gitdb-4.0.5 gitpython-3.1.2 humanize-2.4.0 pandas-1.0.3 parse-1.15.0 pint-0.11 pytz-2020.1 smmap-3.0.4 tzlocal-2.1 utool
Obtaining file:///ibeis/ubelt
    ERROR: Command errored out with exit status 1:
     command: /virtualenv/env3/bin/python3 -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/ibeis/ubelt/setup.py'"'"'; __file__='"'"'/ibeis/ubelt/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' egg_info --egg-base /tmp/pip-pip-egg-info-71gfnkl9
         cwd: /ibeis/ubelt/
    Complete output (9 lines):
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/ibeis/ubelt/setup.py", line 188, in <module>
        long_description=parse_description(),
      File "/ibeis/ubelt/setup.py", line 51, in parse_description
        text = f.read()
      File "/virtualenv/env3/lib/python3.6/encodings/ascii.py", line 26, in decode
        return codecs.ascii_decode(input, self.errors)[0]
    UnicodeDecodeError: 'ascii' codec can't decode byte 0xc2 in position 10495: ordinal not in range(128)
    ----------------------------------------
ERROR: Command errored out with exit status 1: python setup.py egg_info Check the logs for full command output.
The command '/bin/sh -c cd /ibeis/utool  && /virtualenv/env3/bin/pip install -e .  && cd /ibeis/ubelt  && /virtualenv/env3/bin/pip install -e .' returned a non-zero code: 1
```

**Comment by karenc** [comment]
More errors while building the provision docker image:

```
Step 12/13 : RUN . /virtualenv/env3/bin/activate  && cd /ibeis/flann  && /bin/bash unix_build.sh  && cd /ibeis/hesaff  && /bin/bash unix_build.sh  && cd /ibeis/lightnet  && /virtualenv/env3/bin/pip install -r requirements.txt  && /virtualenv/env3/bin/pip install -r develop.txt  && cd /ibeis/ibeis-flukematch-module  && /bin/bash unix_build.sh  && cd /ibeis/ibeis-curvrank-module  && /bin/bash unix_build.sh  && cd /ibeis/pydarknet  && /bin/bash unix_build.sh  && cd /ibeis/pyrf  && /bin/bash unix_build.sh  && cd /ibeis/vtool  && /bin/bash unix_build.sh
 ---> Running in dcb3c2b1ede6
PYEXE              = /virtualenv/env3/bin/python
PYTHON_EXECUTABLE  = /virtualenv/env3/bin/python
LOCAL_PREFIX       = /virtualenv/env3
_SUDO              =
CMake Error at CMakeLists.txt:1 (cmake_minimum_required):
  CMake 3.13.0 or higher is required.  You are running version 3.10.2


-- Configuring incomplete, errors occurred!
make: *** No targets specified and no makefile found.  Stop.
/ibeis/flann
unix_build.sh: line 71: cd: /ibeis/flann/src/python: No such file or directory
python: can't open file 'setup.py': [Errno 2] No such file or directory
ERROR: /ibeis/flann/src/python is not a valid editable requirement. It should either be a path to a local project or a VCS URL (beginning with svn+, git+, hg+, or bzr+).
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ModuleNotFoundError: No module named 'pyflann'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ModuleNotFoundError: No module named 'pyflann'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ModuleNotFoundError: No module named 'pyflann'
/bin/bash: unix_build.sh: No such file or directory
```

**Comment by karenc** [comment]
After fixing the cmake version error, more errors occurred when building the provision docker image:

```
Step 12/13 : RUN . /virtualenv/env3/bin/activate  && cd /ibeis/flann  && /bin/bash unix_build.sh  && cd /ibeis/hesaff  && /bin/bash unix_build.sh  && cd /ibeis/lightnet  && /virtualenv/env3/bin/pip install -r requirements.txt  && /virtualenv/env3/bin/pip install -r develop.txt  && cd /ibeis/ibeis-flukematch-module  && /bin/bash unix_build.sh  && cd /ibeis/ibeis-curvrank-module  && /bin/bash unix_build.sh  && cd /ibeis/pydarknet  && /bin/bash unix_build.sh  && cd /ibeis/pyrf  && /bin/bash unix_build.sh  && cd /ibeis/vtool  && /bin/bash unix_build.sh
 ---> Running in 58cdce758823
...
-- Found PkgConfig: /usr/bin/pkg-config (found version "0.29.1")
-- Checking for module 'liblz4'
--   No package 'liblz4' found
CMake Error at /usr/local/share/cmake-3.17/Modules/FindPkgConfig.cmake:467 (message):
  A required package was not found
Call Stack (most recent call first):
  /usr/local/share/cmake-3.17/Modules/FindPkgConfig.cmake:647 (_pkg_check_modules_internal)
  CMakeLists.txt:156 (pkg_check_modules)


-- Configuring incomplete, errors occurred!
```

`liblz4-dev` seems to be a requirement for flann now so needs to be installed.

```
+ pip install -e /ibeis/flann/
Obtaining file:///ibeis/flann
    ERROR: Command errored out with exit status 1:
     command: /virtualenv/env3/bin/python3 -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/ibeis/flann/setup.py'"'"'; __file__='"'"'/ibeis/flann/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' egg_info --egg-base /tmp/pip-pip-egg-info-dffc8kol
         cwd: /ibeis/flann/
    Complete output (8 lines):
    running egg_info
    creating /tmp/pip-pip-egg-info-dffc8kol/pyflann_ibeis.egg-info
    writing /tmp/pip-pip-egg-info-dffc8kol/pyflann_ibeis.egg-info/PKG-INFO
    writing dependency_links to /tmp/pip-pip-egg-info-dffc8kol/pyflann_ibeis.egg-info/dependency_links.txt
    writing requirements to /tmp/pip-pip-egg-info-dffc8kol/pyflann_ibeis.egg-info/requires.txt
    writing top-level names to /tmp/pip-pip-egg-info-dffc8kol/pyflann_ibeis.egg-info/top_level.txt
    writing manifest file '/tmp/pip-pip-egg-info-dffc8kol/pyflann_ibeis.egg-info/SOURCES.txt'
    error: package directory 'pyflann_ibeis/lib' does not exist
    ----------------------------------------
ERROR: Command errored out with exit status 1: python setup.py egg_info Check the logs for full command output.
```

The [unix_build.sh](https://github.com/WildbookOrg/flann/blob/17de08d196c9a89c4002d559a20cef10596235be/unix_build.sh) in flann is really outdated so that needs to be fixed first.

**Comment by bluemellophone** [comment]
@karenc, Let me know when this PR is ready and I'll take a look.  I'm not going to follow the activity here unless there is a question or an issue you cannot get past.

**Comment by karenc** [comment]
I plan to rebase the commits onto master / master-surgery instead. It might fix some of the temporary install due to renaming.

**Comment by karenc** [comment]
This branch is now based on `develop`.

Commented out the `super_setup.py` lines in the docker files because the repos are different (e.g. `guitool` vs `guitool_ibeis`).

File: devops/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+ARG BRANCH=next
+ARG AZURE_DEVOPS_CACHEBUSTER=0
+
+RUN echo "ARGS BRANCH=${BRANCH}"
+RUN echo "ARGS AZURE_DEVOPS_CACHEBUSTER=${AZURE_DEVOPS_CACHEBUSTER}"
+
+RUN git config --global user.email "dev@wildme.org" \
+ && git config --global user.name "Wild Me Development Nightly Build (Azure CI)"
+
+# TODO Temporarily disabled: Checkout latest version of all other repos
```
Comment (bluemellophone) [inline devops/Dockerfile:14]: Let's re-enable this

File: devops/provision/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+ && git clone --branch next https://github.com/WildbookOrg/flann                       /ibeis/flann \
+ && git clone --branch next https://github.com/WildbookOrg/guitool                     /ibeis/guitool \
+ && git clone --branch next https://github.com/WildbookOrg/hesaff                      /ibeis/hesaff \
+ && git clone --branch next https://github.com/WildbookOrg/ibeis-flukematch-module     /ibeis/ibeis-flukematch-module \
+ && git clone --branch next https://github.com/WildbookOrg/ibeis-curvrank-module       /ibeis/ibeis-curvrank-module \
+ && git clone --branch next https://github.com/WildbookOrg/ibeis-deepsense-module      /ibeis/ibeis-deepsense-module \
+ && git clone --branch next https://github.com/WildbookOrg/ibeis-finfindr-module       /ibeis/ibeis-finfindr-module \
+ && git clone --branch next https://github.com/WildbookOrg/ibeis-kaggle7-module        /ibeis/ibeis-kaggle7-module \
+ && git clone --branch next https://github.com/WildbookOrg/ibeis-2d-orientation-module /ibeis/ibeis-2d-orientation-module \
+ && git clone --branch master https://github.com/WildbookOrg/ibeis                     /ibeis/ibeis \
```
Comment (bluemellophone) [inline devops/provision/Dockerfile:25]: Let's use next for the immediate term

**Pull request merged**